impeller/renderer/backend/gles/gpu_tracer_gles.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/renderer/backend/gles/gpu_tracer_gles.h"
+#include <thread>
+#include "fml/trace_event.h"
+
+namespace impeller {
+
+GPUTracerGLES::GPUTracerGLES(const ProcTableGLES& gl) {
+#ifdef IMPELLER_DEBUG
+  auto desc = gl.GetDescription();
+  enabled_ = desc->HasExtension("GL_EXT_disjoint_timer_query");
+#endif  // IMPELLER_DEBUG
+}
+
+void GPUTracerGLES::MarkFrameStart(const ProcTableGLES& gl) {
+  if (!enabled_ || has_started_frame_ ||
+      std::this_thread::get_id() != raster_thread_) {
+    return;
+  }
+
+  // At the beginning of a frame, check the status of all pending
+  // previous queries.
+  ProcessQueries(gl);
+
+  uint32_t query = 0;
+  gl.GenQueriesEXT(1, &query);
+  if (query == 0) {
+    return;
+  }
+
+  has_started_frame_ = true;
+
+  FML_DCHECK(!active_frame_.has_value());
+  FML_DCHECK(query != 0);
+  active_frame_ = query;
+  gl.BeginQueryEXT(GL_TIME_ELAPSED_EXT, query);
+}
+
+void GPUTracerGLES::RecordRasterThread() {
+  raster_thread_ = std::this_thread::get_id();
+}
+
+void GPUTracerGLES::ProcessQueries(const ProcTableGLES& gl) {
+  if (pending_traces_.empty()) {
+    return;
+  }
+
+  // For reasons unknown to me, querying the state of more than
+  // on query object per frame causes crashes on a Pixel 6 pro.
+  // It does not crash on an S10.
```

Comment:
Where exactly is the crash happening?

If the tracer can only consume one pending query per `MarkFrameStart` call, then that means the pending traces queue will grow unbounded each time a query result is unavailable at the next `MarkFrameStart`.